homeassistant/components/ondilo_ico/sensor.py

```diff
@@ -0,0 +1,185 @@
+        DEVICE_CLASS_SIGNAL_STRENGTH,
+    ],
+    "salt": ["Salt", "mg/L", "mdi:pool", None],
+}
+
+SCAN_INTERVAL = timedelta(hours=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Ondilo ICO sensors."""
+
+    api = hass.data[DOMAIN][entry.entry_id]
+
+    def get_all_pool_data(pool):
+        """Add pool details and last measures to pool data."""
+        pool["ICO"] = api.get_ICO_details(pool["id"])
+        pool["sensors"] = api.get_last_pool_measures(pool["id"])
+
+        return pool
+
+    async def async_update_data():
+        """Fetch data from API endpoint.
+
+        This is the place to pre-process the data to lookup tables
+        so entities can quickly look up their data.
+        """
+        try:
+            pools = await hass.async_add_executor_job(api.get_pools)
+
+            return await asyncio.gather(
+                *[
+                    hass.async_add_executor_job(get_all_pool_data, pool)
+                    for pool in pools
+                ]
+            )
+
+        except OndiloError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="sensor",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=SCAN_INTERVAL,
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    entities = []
+    for poolidx, pool in enumerate(coordinator.data):
+        for sensor_idx, sensor in enumerate(pool["sensors"]):
+            if SENSOR_TYPES.get(sensor["data_type"], None) is not None:
+                entities.append(OndiloICO(coordinator, poolidx, sensor_idx))
+
+    async_add_entities(entities)
+
+
+class OndiloICO(CoordinatorEntity):
+    """Representation of a Sensor."""
+
+    def __init__(
+        self, coordinator: DataUpdateCoordinator, poolidx: int, sensor_idx: int
+    ):
+        """Initialize sensor entity with data from coordinator."""
+        super().__init__(coordinator)
+
+        self._poolid = self.coordinator.data[poolidx]["id"]
+
+        pooldata = self._pooldata()
+        self._data_type = pooldata["sensors"][sensor_idx]["data_type"]
+        self._unique_id = f"{pooldata['ICO']['serial_number']}-{self._data_type}"
+        self._device_name = pooldata["name"]
+        self._name = f"{self._device_name} {SENSOR_TYPES[self._data_type][0]}"
+        self._device_class = SENSOR_TYPES[self._data_type][3]
+        self._icon = SENSOR_TYPES[self._data_type][2]
+        self._unit = SENSOR_TYPES[self._data_type][1]
+
+    def _pooldata(self):
+        """Get pool data dict."""
+        return next(
+            (pool for pool in self.coordinator.data if pool["id"] == self._poolid),
```

Comment:
I could see doing this work up front in `async_update_data` as another approach to consider in a future PR, though i realize the N count is probably low here.

Reply:
Correct. Still, since I was also worried of that, I try to avoid multiple calls to it (and `_devdata`) in the same function. That being said, you're right, I think most people only have one ICO so it shouldn't be an issue.